content_scripts/vimium_frontend.coffee

```diff
@@ -342,28 +376,37 @@ extend window,
     hintContainingDiv = DomUtils.addElementList(hints,
       { id: "vimiumInputMarkerContainer", className: "vimiumReset" })
 
-    handlerStack.push keydown: (event) ->
-      if event.keyCode == KeyboardUtils.keyCodes.tab
-        hints[selectedInputIndex].classList.remove 'internalVimiumSelectedInputHint'
-        if event.shiftKey
-          if --selectedInputIndex == -1
-            selectedInputIndex = hints.length - 1
-        else
-          if ++selectedInputIndex == hints.length
-            selectedInputIndex = 0
-        hints[selectedInputIndex].classList.add 'internalVimiumSelectedInputHint'
+    new class FocusSelector extends InsertModeBlocker
+      constructor: ->
+        super null,
+          name: "focus-selector"
+          badge: "?"
+          keydown: (event) =>
+            if event.keyCode == KeyboardUtils.keyCodes.tab
+              hints[selectedInputIndex].classList.remove 'internalVimiumSelectedInputHint'
+              selectedInputIndex += hints.length + (if event.shiftKey then -1 else 1)
+              selectedInputIndex %= hints.length
+              hints[selectedInputIndex].classList.add 'internalVimiumSelectedInputHint'
+              visibleInputs[selectedInputIndex].element.focus()
+              false
+            else unless event.keyCode == KeyboardUtils.keyCodes.shiftKey
+              @exit()
+              @continueBubbling
+
         visibleInputs[selectedInputIndex].element.focus()
-      else unless event.keyCode == KeyboardUtils.keyCodes.shiftKey
-        DomUtils.removeElement hintContainingDiv
-        @remove()
-        return true
+        @exit() if visibleInputs.length == 1
 
-      false
+      exit: ->
+        DomUtils.removeElement hintContainingDiv
+        visibleInputs[selectedInputIndex].element.focus()
+        new InsertMode visibleInputs[selectedInputIndex].element
+        super()
 
 # Decide whether this keyChar should be passed to the underlying page.
 # Keystrokes are *never* considered passKeys if the keyQueue is not empty.  So, for example, if 't' is a
 # passKey, then 'gt' and '99t' will neverthless be handled by vimium.
 isPassKey = ( keyChar ) ->
+  return false # Diabled.
```

Comment:
> Perhaps add a TODO so we remember to delete this code.

Actually, there's a quite a lot of unused code now in the front end.  Most of it's just sitting there.  In a couple of places (like this), I disabled things to *make sure* they're not being used.

I'll come back and strip out the unused code later, once we get this and #1401 sorted.